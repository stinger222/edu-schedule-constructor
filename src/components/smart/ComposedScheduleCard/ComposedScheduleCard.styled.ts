import styled from "styled-components"

export const StyledComposedSchedule = styled.div`
	width: 37em;
	height: fit-content;
	padding: 0 0.3em 1em 0.3em;

	border-radius: 1.3em;
	border: 0.08em solid #00000035;
	box-shadow: ${({theme}) => theme.boxShadows.primary};
	background: ${({theme}) => theme.backgrounds.secondary};
	
	& > header {
		line-height: 1em;
		font-size: 1.9em;
		font-weight: 400;
		letter-spacing: 0.05em;
		padding-block: 0.9em;
		text-align: center;
	}

	& .days {
		display: flex;
		flex-wrap: wrap;
		justify-content: center;
		gap: 0.6em;
		row-gap: 0.7em;
	}

	& .days > * {
		flex-basis: calc(33% - 0.4em);
	}
`