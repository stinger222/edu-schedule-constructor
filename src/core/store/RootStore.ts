import { configure } from "mobx"
import { IComposedSchedulesStore, ILessonsStore, IClassSchedulesStore, IUIStore } from "./../types/store"
import ComposedSchedulesStore from "./ComposedSchedulesStore"
import ClassSchedulesStore from "./ClassSchedulesStore"
import LessonsStore from "./LessonsStore"
import UIStore from "./UIStore"
class RootStore {
	uiStore: IUIStore
	lessonsStore: ILessonsStore
	classSchedulesStore: IClassSchedulesStore
	composedSchedulesStore: IComposedSchedulesStore

	constructor() {
		this.uiStore = new UIStore()
		this.lessonsStore = new LessonsStore()
		this.classSchedulesStore = new ClassSchedulesStore()
		this.composedSchedulesStore = new ComposedSchedulesStore()
	}
}

configure({
    enforceActions: "never"
})

export default RootStore