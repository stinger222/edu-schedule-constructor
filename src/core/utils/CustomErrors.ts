export class EmptyDay extends Error {
  readonly name = "EmptyDay"

  constructor() {
    super()
  }
}

export class Warning extends Error {
  readonly name = "Warning"

  constructor(message: string) {
    super(message)
  }
}
