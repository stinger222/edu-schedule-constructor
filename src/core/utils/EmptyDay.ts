class EmptyDay extends Error {
  name = "EmptyDay"

  constructor(message: string) {
    super(message)
  }
}

export default EmptyDay
