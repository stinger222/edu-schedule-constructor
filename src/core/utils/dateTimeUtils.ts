import { Cases, weekDates } from "../types/types"
import { formatNumber } from "./stringUtils"


interface IData {
  ru: {
    short: readonly string[],
    full: {
      nominative: readonly string[],
      accusative: readonly string[]
    }
  },
  en: {
    short: readonly string[],
    full: {
      nominative: readonly string[],
      accusative: readonly string[]
    }
  }
}
export class WeekDays {

  private static readonly data: IData = {
    ru: {
      short: ["Пн", "Вт", "Ср", "Чт", "Пт", "Сб", "Вс"],
      full: {
        nominative: [
          "Понедельник", "Вторник", "Среда", "Четверг", "Пятница", "Суббота", "Воскресенье"
        ],
        accusative: [
          "Понедельник", "Вторник", "Среду", "Четверг", "Пятницу", "Субботу", "Воскресенье"
        ]
      }
    },
    en: {
      short: ["M", "T", "W", "T", "F", "S", "S"],
      full: {
        nominative: [
          "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"
        ],
        accusative: [
          "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"
        ]
      }
    }
  }

	static getShort(lang: "ru" | "en"): ReadonlyArray<string> {
		return this.data[lang].short as ReadonlyArray<string>
	}

	static getFull(lang: "ru" | "en", caseType: Cases = Cases.Nominative, toLowerCase: boolean = false): ReadonlyArray<string> {
		const result: ReadonlyArray<string> = this.data[lang].full[caseType]
		return toLowerCase ? result.map(i => i.toLowerCase()) : result
	}
}

export const getCurrentWeekDates = (): weekDates => {
	const date = new Date()
	date.setDate(date.getDate() - (date.getDay() === 0 ? 7 : date.getDay())) // now it's current week sunday date:

	const result = Array.from({ length: 7 }, () => {
		date.setDate(date.getDate() + 1) // and now it's monday
		return formatNumber(date.getDate())
	})

	return result as weekDates
}
