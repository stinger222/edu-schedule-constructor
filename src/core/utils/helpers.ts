import { weekDates, Replacements, Cases } from './../types/types';

export const getCurrentWeekDates = (): weekDates => {
	const result = []
	const date = new Date()
	date.setDate(date.getDate() - (date.getDay() === 0 ? 7 : date.getDay())) // now it's current week sunday date
	for (let i = 0; i < 7; i++) {
		date.setDate(date.getDate() + 1) // and now it's monday
		result.push(date.getDate())
	}

	return formatNumbers(result) as weekDates
}

export const formatNumbers = (numbers: number[]): string[] => {
	return numbers.map(number => formatNumber(number))
}

export function replaceBlankProps<P extends {[key: string]: any}> (props: P , replacements: Replacements<P>): P {
	const result: any = {}
	for (const [key, value] of Object.entries(props)) {
		if (typeof value === 'string') {
			if (value.trim() == '') {
				result[key] = replacements[key as keyof typeof replacements]
			} else {
				result[key] = value
			}
		} else {
			result[key] = value
		}
	}
	return result as P
}

export const formatNumber = (number: number): string => {
	if (isNaN(number)) return "??"
	
	return Math.abs(number).toLocaleString('en-US', {
    minimumIntegerDigits: 2,
    useGrouping: false
  })
}

export const formatTimeString = (timeString: string): string => {
	let [hours, minutes] = timeString.split(':')

	return `${formatNumber(parseInt(hours))}:${formatNumber(parseInt(minutes))}`
}

export class WeekDays {
	private static _week = {
		short: ['Пн', 'Вт', 'Ср', 'Чт', 'Пт', 'Сб', 'Вс'],
		full: {
			nominative: [
				'Понедельник', 'Вторник', 'Среда', 'Четверг', 'Пятница', 'Суббота', 'Воскресенье'
			],
			accusative: [
				'Понедельник', 'Вторник', 'Среду', 'Четверг', 'Пятницу', 'Субботу', 'Воскресенье'
			]
		} as Record<Cases, string[]>
	}

	static getShort() {
		return this._week.short
	}

	static getFull(_case: Cases = Cases.Nominative, lowerCase: boolean = false) {
		let result = this._week.full[_case]
		return lowerCase ? result.map(i => i.toLowerCase()) : result
	}
}