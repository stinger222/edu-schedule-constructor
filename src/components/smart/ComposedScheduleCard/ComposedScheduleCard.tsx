import { useContext } from "react"
import { StoreContext } from "../../.."
import { IComposedDay } from "../../../core/types/types"
import { StyledComposedSchedule } from "./ComposedScheduleCard.styled"

import Day from "./Day/Day"
import Switch from "../../ui/Switch/Switch"

interface IProps {
	name: string,
	days: IComposedDay[]
  uid: string
}

const ComposedSchedule: React.FC<IProps> = ({ name, days, uid }) => {

  const { composedSchedulesStore } = useContext(StoreContext)
  const thisIsActive = composedSchedulesStore.activeScheduleUid === uid

  const handleActivation = () => {
    composedSchedulesStore.activateSchedule(uid)
  }

  return (
		<StyledComposedSchedule className="composed-schedule">
			<header>
        <div className="title">{ name }</div>
        <Switch type="checkbox" onChange={handleActivation} checked={thisIsActive} />
      </header>
      <div className="days">
				{
					new Array(5).fill(0).map((_, index) => {
						const day = days[index]
						return (
							<Day
								dayIndex={index}
								day={day}
								key={index}
							/>
						)
					})
				}
			</div>
		</StyledComposedSchedule>
	)
}

export default ComposedSchedule
