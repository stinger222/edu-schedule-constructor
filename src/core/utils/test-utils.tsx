import React, { ReactElement } from "react"
import { HashRouter } from "react-router-dom"
import { render, RenderOptions } from "@testing-library/react"

import RootStore from "../store/RootStore"


import { ThemeProvider } from "styled-components"
import { DarkTheme } from "../themes/Dark"

const AllTheProviders: React.FC<{children: ReactElement}> = ({children}) => {

	const rootStore = new RootStore()
	const StoreContext = React.createContext<typeof rootStore>(rootStore)

	return (
		<ThemeProvider theme={DarkTheme}>
			<StoreContext.Provider value={rootStore}>
				<HashRouter>
					{children}
				</HashRouter>
			</StoreContext.Provider>
		</ThemeProvider>
	)
}

const customRender = (ui: React.ReactElement, options?: Omit<RenderOptions, "queries">) => {
	return render(ui, { wrapper: AllTheProviders, ...options})
}

export * from "@testing-library/react"
export {customRender as render}