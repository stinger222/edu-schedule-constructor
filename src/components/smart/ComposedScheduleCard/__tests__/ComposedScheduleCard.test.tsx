import { renderHook } from "@testing-library/react"
import { useContext } from "react"
import { StoreContext } from "../../../.."
import { render } from "../../../../core/utils/test-utils"
import RootStore from "../../../../core/store/RootStore"
import ComposedScheduleCard from "../ComposedScheduleCard"

let rootStore: RootStore
describe("Testing ComposedScheduleCard component", () => {
  beforeAll(() => {
    rootStore = renderHook(() => useContext(StoreContext)).result.current
    rootStore.ringsSchedulesStore.addSchedule({
      name: "Test Rings Schedule (4 lessons described)",
      rings: [
        {start: "08:30", end: "10:00"},
        {start: "10:30", end: "12:00"},
        {start: "12:30", end: "14:00"},
        {start: "14:30", end: "16:00"}
      ]
    }, "test-rings-schedule")
  })

  it("Properly renders schedule card", () => {
    const screen = render(
      <ComposedScheduleCard
        name="Test Schedule #1"
        days={[
          {ringsScheduleId: "test-rings-schedule", lessonIds: ["1", "2", "3", "4"]},
          {ringsScheduleId: "test-rings-schedule", lessonIds: ["hidden", "2", "3", "4"]},
          {ringsScheduleId: "test-rings-schedule", lessonIds: ["hidden", "hidden", "3", "4"]},
          {ringsScheduleId: "test-rings-schedule", lessonIds: ["hidden"]},
          {ringsScheduleId: "test-rings-schedule", lessonIds: ["hidden", "hidden", "hidden", "hidden"]}
        ]}
        uid="test-composed-schedule-id-1"
      />
    )
    expect(screen).toMatchSnapshot()
  })
})