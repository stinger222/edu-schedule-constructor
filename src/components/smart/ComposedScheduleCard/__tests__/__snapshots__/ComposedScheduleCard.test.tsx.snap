// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ComposedScheduleCard component Properly renders schedule card 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div />
    <div>
      <div
        class="sc-eDWCr hoVdDn composed-schedule"
      >
        <header>
          <div
            class="title"
          >
            Test Schedule #1
          </div>
          <input
            class="sc-ksBlkl jvriEz switch"
            type="checkbox"
          />
        </header>
        <div
          class="days"
        >
          <div
            class="sc-bqWxrE gTNlmT"
          >
            <header>
              M
            </header>
            <div
              class="card-body"
            >
              <span>
                 
                Lessons
                 
              </span>
              <span>
                4
              </span>
              <span>
                 
                Start
                 
              </span>
              <span>
                08:30
              </span>
              <span>
                 
                End
                 
              </span>
              <span>
                16:00
              </span>
            </div>
          </div>
          <div
            class="sc-bqWxrE gTNlmT"
          >
            <header>
              T
            </header>
            <div
              class="card-body"
            >
              <span>
                 
                Lessons
                 
              </span>
              <span>
                3
              </span>
              <span>
                 
                Start
                 
              </span>
              <span>
                10:30
              </span>
              <span>
                 
                End
                 
              </span>
              <span>
                16:00
              </span>
            </div>
          </div>
          <div
            class="sc-bqWxrE gTNlmT"
          >
            <header>
              W
            </header>
            <div
              class="card-body"
            >
              <span>
                 
                Lessons
                 
              </span>
              <span>
                2
              </span>
              <span>
                 
                Start
                 
              </span>
              <span>
                12:30
              </span>
              <span>
                 
                End
                 
              </span>
              <span>
                16:00
              </span>
            </div>
          </div>
          <div
            class="sc-bqWxrE gTNlmT"
          >
            <header>
              T
            </header>
            <div
              class="card-body"
            >
              <span>
                 
                Lessons
                 
              </span>
              <span>
                0
              </span>
              <span>
                 
                Start
                 
              </span>
              <span>
                -:-
              </span>
              <span>
                 
                End
                 
              </span>
              <span>
                -:-
              </span>
            </div>
          </div>
          <div
            class="sc-bqWxrE gTNlmT"
          >
            <header>
              F
            </header>
            <div
              class="card-body"
            >
              <span>
                 
                Lessons
                 
              </span>
              <span>
                0
              </span>
              <span>
                 
                Start
                 
              </span>
              <span>
                -:-
              </span>
              <span>
                 
                End
                 
              </span>
              <span>
                -:-
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-eDWCr hoVdDn composed-schedule"
    >
      <header>
        <div
          class="title"
        >
          Test Schedule #1
        </div>
        <input
          class="sc-ksBlkl jvriEz switch"
          type="checkbox"
        />
      </header>
      <div
        class="days"
      >
        <div
          class="sc-bqWxrE gTNlmT"
        >
          <header>
            M
          </header>
          <div
            class="card-body"
          >
            <span>
               
              Lessons
               
            </span>
            <span>
              4
            </span>
            <span>
               
              Start
               
            </span>
            <span>
              08:30
            </span>
            <span>
               
              End
               
            </span>
            <span>
              16:00
            </span>
          </div>
        </div>
        <div
          class="sc-bqWxrE gTNlmT"
        >
          <header>
            T
          </header>
          <div
            class="card-body"
          >
            <span>
               
              Lessons
               
            </span>
            <span>
              3
            </span>
            <span>
               
              Start
               
            </span>
            <span>
              10:30
            </span>
            <span>
               
              End
               
            </span>
            <span>
              16:00
            </span>
          </div>
        </div>
        <div
          class="sc-bqWxrE gTNlmT"
        >
          <header>
            W
          </header>
          <div
            class="card-body"
          >
            <span>
               
              Lessons
               
            </span>
            <span>
              2
            </span>
            <span>
               
              Start
               
            </span>
            <span>
              12:30
            </span>
            <span>
               
              End
               
            </span>
            <span>
              16:00
            </span>
          </div>
        </div>
        <div
          class="sc-bqWxrE gTNlmT"
        >
          <header>
            T
          </header>
          <div
            class="card-body"
          >
            <span>
               
              Lessons
               
            </span>
            <span>
              0
            </span>
            <span>
               
              Start
               
            </span>
            <span>
              -:-
            </span>
            <span>
               
              End
               
            </span>
            <span>
              -:-
            </span>
          </div>
        </div>
        <div
          class="sc-bqWxrE gTNlmT"
        >
          <header>
            F
          </header>
          <div
            class="card-body"
          >
            <span>
               
              Lessons
               
            </span>
            <span>
              0
            </span>
            <span>
               
              Start
               
            </span>
            <span>
              -:-
            </span>
            <span>
               
              End
               
            </span>
            <span>
              -:-
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
